extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java

```diff
@@ -159,14 +159,30 @@ public QueryDefinition makeQueryDefinition(
         clusterBy.getColumns()
     );
 
-    // If there is no limit spec, apply the final shuffling here itself. This will ensure partition sizes etc are respected.
-    // If there is a limit spec, it would write everything into one partition anyway, so the final shuffling should be
-    // applied after that.
+    ShuffleSpec scanShuffleSpec;
+    if (!hasLimitOrOffset) {
+      // If there is no limit spec, apply the final shuffling here itself. This will ensure partition sizes etc are respected.
+      scanShuffleSpec = finalShuffleSpec;
+    } else {
+      // If there is a limit spec, check if there are any non-boost columns to sort in.
+      boolean requiresSort = clusterByColumns.stream()
```

Comment:
Do we need to see if we are sorting on the boost column? Isn't that automatically added by the ScanQueryKit et al?

Reply:
We have a if-branch for window functions above this bit of code, which adds its own cluster bys, which are not dependent on the orderBy.